### PR TITLE
[CDAP-18955] Add support for remote program run dispatch tasks

### PR DESCRIPTION
**Pull request by masoud-io**
This PR modifies system worker pods and add support for accepting remote program run dispatch tasks [PR](https://github.com/cdapio/cdap/pull/14090).

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/guice/ImpersonatedTwillRunnerService.java [hunk 1]
```diff
@@ -44,14 +44,14 @@
...
+public final class ImpersonatedTwillRunnerService implements TwillRunnerService {
 
   private final Configuration hConf;
   private final TwillRunnerService delegate;
   private final Impersonator impersonator;
   private final TokenSecureStoreRenewer secureStoreRenewer;
 
-  ImpersonatedTwillRunnerService(Configuration hConf, TwillRunnerService delegate, Impersonator impersonator,
+  public ImpersonatedTwillRunnerService(Configuration hConf, TwillRunnerService delegate, Impersonator impersonator,
                                  TokenSecureStoreRenewer secureStoreRenewer) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/guice/ImpersonatedTwillRunnerService.java]: indentation

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/RemoteTwillModule.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import com.google.inject.Scopes;
+import io.cdap.cdap.app.guice.ImpersonatedTwillRunnerService;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.security.TokenSecureStoreRenewer;
+import io.cdap.cdap.security.impersonation.Impersonator;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.twill.api.TwillRunner;
+import org.apache.twill.api.TwillRunnerService;
+
+public class RemoteTwillModule extends AbstractModule {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/RemoteTwillModule.java:36]: Can you add a javadoc with some explanation of what this is used for

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerHttpHandlerInternal.java [hunk 3]
```diff
@@ -90,8 +92,10 @@
 
   private final MetricsCollectionService metricsCollectionService;
 
-  public SystemWorkerHttpHandlerInternal(CConfiguration cConf, MetricsCollectionService metricsCollectionService) {
-    this.runnableTaskLauncher = new RunnableTaskLauncher(cConf);
+  public SystemWorkerHttpHandlerInternal(CConfiguration cConf, MetricsCollectionService metricsCollectionService,
+    Injector injector) {
+    this.runnableTaskLauncher = new RunnableTaskLauncher(cConf, injector);
+    this.injector = injector;
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerHttpHandlerInternal.java]: seems unused?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerService.java [hunk 4]
```diff
@@ -64,9 +71,14 @@
   SystemWorkerService(CConfiguration cConf,
       SConfiguration sConf,
       DiscoveryService discoveryService,
-      ArtifactManagerFactory artifactManagerFactory, MetricsCollectionService metricsCollectionService) {
+      ArtifactManagerFactory artifactManagerFactory, MetricsCollectionService metricsCollectionService,
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerService.java]: indentation

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerHttpHandlerInternal.java [hunk 5]
```diff
@@ -90,8 +92,10 @@
 
   private final MetricsCollectionService metricsCollectionService;
 
-  public SystemWorkerHttpHandlerInternal(CConfiguration cConf, MetricsCollectionService metricsCollectionService) {
-    this.runnableTaskLauncher = new RunnableTaskLauncher(cConf);
+  public SystemWorkerHttpHandlerInternal(CConfiguration cConf, MetricsCollectionService metricsCollectionService,
+    Injector injector) {
+    this.runnableTaskLauncher = new RunnableTaskLauncher(cConf, injector);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerHttpHandlerInternal.java]: it's kind of weird to pass an injector around, any reason we can't pass the RunnableTaskLauncher in here?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java [hunk 6]
```diff
@@ -81,26 +114,75 @@
...
     modules.add(new MetricsClientRuntimeModule().getDistributedModules());
 
+    modules.addAll(Arrays.asList(
+      // Always use local table implementations, which use LevelDB.
+      // In K8s, there won't be HBase and the cdap-site should be set to use SQL store for StructuredTable.
+      new DataSetServiceModules().getStandaloneModules(),
+      // The Dataset set modules are only needed to satisfy dependency injection
+      new DataSetsModules().getStandaloneModules(),
+      new MetricsStoreModule(),
+      new MessagingClientModule(),
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java:157]: MessagingClientModule is already added above

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java [hunk 7]
```diff
@@ -81,26 +114,75 @@
...
+      new ExploreClientModule(),
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java:158]: is this needed?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java [hunk 8]
```diff
@@ -81,26 +114,75 @@
...
+          @Override
+          protected void configure() {
+            bind(StorageProviderNamespaceAdmin.class).to(
+              LocalStorageProviderNamespaceAdmin.class);
+          }
+        }),
+      new ProgramRunnerRuntimeModule().getDistributedModules(true),
+      new MonitorHandlerModule(false),
+      new SecureStoreServerModule(),
+      new OperationalStatsModule(),
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java:172]: is this needed?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java (hunk 6 above)
Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java:157]: Thanks for catching this.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java (hunk 8 above)
Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java:172]: We'll test this to make sure it's necessary once PR1 & PR2 are merged.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java (hunk 7 above)
Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java:158]: We'll test this to make sure it's necessary once PR1 & PR2 are merged.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerHttpHandlerInternal.java (hunk 5 above)
Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerHttpHandlerInternal.java]: The need for passing injector was discussed offline.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/RemoteTwillModule.java (hunk 2 above)
Comment (adityagupta94) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/RemoteTwillModule.java:36]: Added

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/RunnableTaskLauncher.java [hunk 9]
```diff
@@ -27,10 +27,14 @@
...
 public class RunnableTaskLauncher {
-  private final CConfiguration cConf;
+  private final Injector injector;
 
   public RunnableTaskLauncher(CConfiguration cConf) {
-    this.cConf = cConf;
+    injector = Guice.createInjector(new RunnableTaskModule(cConf));
+  }
+
+  public RunnableTaskLauncher(Injector injector) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/RunnableTaskLauncher.java:40]: can you add some comments on when this is used and how it is different than the other constructor?

**Review by albertshau — APPROVED** [review]
lgtm assuming we'll revisit whether some of the guice modules are needed in a later PR.

**Pull request merged**